### ISSUE TITLE
Update value of fundamental constants

### ISSUE THREAD

**Issue by DiscreteLogarithm** [lammps/lammps#2324]
**Description**

In 2018 and 2019 four fundamental constants (Planck constant, Boltzmann constant, Avogadro constant, and elementary charge) have been redefined to have exact constant values and four of the SI base units (kilogram, ampere, kelvin, and mole) have been redefined in terms of these constants. The values used by LAMMPS differ from the new values (usually in the 6th significant digit so the adjustments are slight). Many conversion factors also depend on these constants. Hopefully all of them are in `update.cpp`. Since the redefinition these constants are no longer empirical, so hopefully they will actually be constant this time (unless redefined).

**Comment by akohlmey** [comment]
This is not as trivial an issue as you might think. There are multiple issues at hand and for each of them there is no simple/fast solution.

1) Changing constants will break backward compatibility of results and all regression and many unit tests. Since the differences in the units are usually much smaller than the systematic errors of the models employed and the statistical errors of statistical mechanical analysis, there is not a lot gained from the change by itself, but it will cause a lot of work for the maintainer and also will worry a lot of people that compare results between different versions of LAMMPS.

2) The various conversion factors in LAMMPS are effectively "ad hoc" and computed with limited precision and not computed to full double precision from a consistent set of fundamental constants. So not only would one need to update the fundamental constants, but also update the derived constants to make a consistent set that can be cleanly interconverted with minimal loss of precision.  The OpenKIM folks have started doing something like that in their `src/KIM/kim_units.cpp` file.

3) Not everybody is aware of the general conversion factors and some people have domain specific units and conversion factors. Those would have to be updated as well, and it is not always easy to determine where some numerical constants come from or what combination of constants and factors they represent.  **If** a general change is made, this would have to be done consistently across all of LAMMPS. Again, a significant effort that would also require reaching out to code contributors to confirm their choices or deriving the constants that code uses properly

4) Some codes make specific choices and there are "hacks" in LAMMPS that (temporarily) switch fundamental constants (e.g. for some newer CHARMM pair styles that change the conversion factor for the coulomb energy to be compatible with the charmm code.

In general, this kind of decision is something that needs a significant discussion and an agreement between the core developers and somebody willing and capable of doing all the required changes and reliably so (it is going to be a *lot* of work for little credit and with the side effect of people complaining that LAMMPS is not working as expected after an upgrade).

**Comment by DiscreteLogarithm** [comment]
I agree that this is a big change that need to be debated and that if a decision is made all of LAMMPS should be changed in a consistent manner. I also agree that the gain is not much. In my opinion users and unit tests should not rely on exact numerical values of the results because of the errors involved (but I know that might be hard to explain to users).

I think that even in the case that the decision is to retain the old values it would still be nice to change the ad hoc conversion factors into expressions in terms of the fundamental constants and store the constants in a single place. But this would still be a breaking change since these values are previously computed with limited precision.

**Comment by akohlmey** [comment]
This is connected to #1377

**Comment by akohlmey** [comment]
@sjplimp this issue has been sitting here for a while without seeing any action. Do we *want* to do something about this issue, or just stick with the values we have?

**Comment by sjplimp** [comment]
Let's see what Aidan thinks - @athomps

**Comment by athomps** [comment]
I understand the elegance and beauty of deriving all the LAMMPS conversion factors from the fundamental constants, whose SI values are now truly constant, a great achievement for science.  However, the current LAMMPS conversion factor values seem to be sufficiently faithful to the SI values that it is not a cause of practical problems.  Certainly, this level of precision is more than adequate for comparing to result of other models or experiments.  The main issue might be comparing with other molecular simulation codes running exactly the same model e.g. OpenKIM.  The best way to handle that is to recognize that there will always be systematic differences between different codes due to different values for conversion factors, and in the rare cases where this matters, these can be accounted for systematically.  Of much greater importance is maintaining internal consistency between different versions of LAMMPS.  Hence I agree with Axel.   This change would add a significant amount of entropy to the LAMMPS-verse, with very little practical benefit.
